# Stock Performance & Volatility Analysis
*Author: Cody Stuerman*

## 1. Project Overview

**Objective:**
To analyze the historical stock performance and volatility of selected companies across different sectors.

**Stocks analyzed:**
YUM, BAC, GE, XOM, PFE

**Time period:**
January 1, 2020 to December 31, 2024

**Key questions:**
- Which stock had the highest average return?
- Which sector showed the most volatility?
- How did different stocks perform during major events like COVID-19

## 2. Data Collection & Preparation

### 2.1 Import Libraries  
We will import essential Python libraries needed for data handling and visualization.

### 2.2 Load Data  
We will load the cleaned stock data from the prepared CSV file into a Pandas Dataframe for analysis

### 2.3 Data Cleaning  
We will:
 - Check for missing values
 - Ensure correct data types.
 - Verify calculated columns ('Daily_Return', and 'Volatility') are present and correct.

In [13]:
import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

In [14]:
df = pd.read_csv(r'C:\Users\Stuer\OneDrive\Professional\Stock_Performance_Volatility_Analysis\data\cleaned\clean_stock_data.csv')
df.head()
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6285 entries, 0 to 6284
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype  
---  ------        --------------  -----  
 0   Date          6285 non-null   object 
 1   Ticker        6285 non-null   object 
 2   Sector        6285 non-null   object 
 3   Open          6285 non-null   float64
 4   High          6285 non-null   float64
 5   Low           6285 non-null   float64
 6   Close         6285 non-null   float64
 7   Adj Close     6285 non-null   float64
 8   Volume        6285 non-null   int64  
 9   Daily_Return  6285 non-null   float64
 10  Volatility    6190 non-null   float64
dtypes: float64(7), int64(1), object(3)
memory usage: 540.2+ KB


After cleaning:
- `Date` column is currently object and will be converted to datetime.
- `Volatility` has 95 missing values (expected due to rolling calculation).
- All other columns have no missing data.
- Calculated columns (`Daily_Return`, `Volatility`) are present.


In [16]:
# Convert date to datetime
df['Date'] = pd.to_datetime(df['Date'])

# Confirm data types after conversion
df.info()

# Check missing values  (should be only in Volatility)
df.isnull().sum()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6285 entries, 0 to 6284
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Date          6285 non-null   datetime64[ns]
 1   Ticker        6285 non-null   object        
 2   Sector        6285 non-null   object        
 3   Open          6285 non-null   float64       
 4   High          6285 non-null   float64       
 5   Low           6285 non-null   float64       
 6   Close         6285 non-null   float64       
 7   Adj Close     6285 non-null   float64       
 8   Volume        6285 non-null   int64         
 9   Daily_Return  6285 non-null   float64       
 10  Volatility    6190 non-null   float64       
dtypes: datetime64[ns](1), float64(7), int64(1), object(2)
memory usage: 540.2+ KB


Date             0
Ticker           0
Sector           0
Open             0
High             0
Low              0
Close            0
Adj Close        0
Volume           0
Daily_Return     0
Volatility      95
dtype: int64

## 3. Exploratory Data Analysis (EDA)

### 3.1 Data Overview  
We will:
- Preview the dataset with .head().
- Confirm columns, data types, and structure with .info().
- Check basic descriptive statistics with .describe().
- Verify the stocks ('Ticker') and sectors included.
- Confirm the date range of the data.

### 3.2 Missing Value Check  
We will:
- Review missing values in each column.
- Focus on confirming that missing data is only in 'Volatility'.
- Determine if any action like drop or fill is needed.

### 3.3 Basic Statistical Summary 
We will:
- Aggregate key metrics such as average daily return and average volatility.
- Group by sector and ticker to compare performance and risk.

### 3.4 Initial Insights  
We will:
- Document initial observations and patterns found during the overview and basic statistics.
- Highlight standout stocks, sectors, or unusual data points.


In [18]:
# Preview the first 5 rows of the dataset
df.head()

Date Ticker                  Sector        Open        High  \
0 2020-01-02    YUM  Consumer Discretionary  100.900002  102.190002   
1 2020-01-03    YUM  Consumer Discretionary  101.419998  102.010002   
2 2020-01-06    YUM  Consumer Discretionary  101.430000  101.800003   
3 2020-01-07    YUM  Consumer Discretionary  101.750000  102.239998   
4 2020-01-08    YUM  Consumer Discretionary  100.470001  102.980003   

          Low       Close  Adj Close   Volume  Daily_Return  Volatility  
0  100.809998  102.169998  92.652504  1369900      0.012587         NaN  
1  100.360001  101.849998  92.362297  1145500      0.004240         NaN  
2  101.059998  101.790001  92.307907  1454100      0.003549         NaN  
3  101.540001  101.970001  92.471130  1388600      0.002162         NaN  
4  100.339996  102.150002  92.634354  1521500      0.016721         NaN

In [19]:
# Check columns, data types, and structure
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 6285 entries, 0 to 6284
Data columns (total 11 columns):
 #   Column        Non-Null Count  Dtype         
---  ------        --------------  -----         
 0   Date          6285 non-null   datetime64[ns]
 1   Ticker        6285 non-null   object        
 2   Sector        6285 non-null   object        
 3   Open          6285 non-null   float64       
 4   High          6285 non-null   float64       
 5   Low           6285 non-null   float64       
 6   Close         6285 non-null   float64       
 7   Adj Close     6285 non-null   float64       
 8   Volume        6285 non-null   int64         
 9   Daily_Return  6285 non-null   float64       
 10  Volatility    6190 non-null   float64       
dtypes: datetime64[ns](1), float64(7), int64(1), object(2)
memory usage: 540.2+ KB


In [20]:
# Review basic descriptive statistics for columns
df.describe()

Date         Open         High          Low  \
count                           6285  6285.000000  6285.000000  6285.000000   
mean   2022-06-30 22:13:27.637231616    70.859646    71.633860    70.083927   
min              2020-01-02 00:00:00    19.260000    19.670000    17.950001   
25%              2021-04-01 00:00:00    36.736244    37.096775    36.200001   
50%              2022-06-30 00:00:00    54.790001    55.639999    53.940800   
75%              2023-09-29 00:00:00   107.230003   108.108543   106.000000   
max              2024-12-30 00:00:00   194.089996   194.800003   192.199997   
std                              NaN    40.093424    40.423717    39.772044   

             Close    Adj Close        Volume  Daily_Return   Volatility  
count  6285.000000  6285.000000  6.285000e+03   6285.000000  6190.000000  
mean     70.862243    66.780416  2.342256e+07      0.000011     0.014758  
min      18.080000    15.967170  3.404000e+05     -0.096048     0.005098  
25%      36.660000    33.069237  6.228788e+06     -0.008519     0.010613  
50%      54.744492    49.048676  1.891660e+07      0.000000     0.013321  
75%     107.139999   101.006538  3.372910e+07      0.008703     0.016804  
max     194.229996   193.553955  2.301539e+08      0.124455     0.053186  
std      40.105432    39.803658  2.174849e+07      0.016003     0.006403

In [21]:
# Verify unique stocks and sectors included in the data
df['Ticker'].unique(), df['Sector'].unique()

(array(['YUM', 'BAC', 'GE', 'XOM', 'PFE'], dtype=object),
 array(['Consumer Discretionary', 'Financials', 'Industrials', 'Energy',
        'Healthcare'], dtype=object))

In [22]:
# Confirm date range of the dataset
df['Date'].min(), df['Date'].max()

(Timestamp('2020-01-02 00:00:00'), Timestamp('2024-12-30 00:00:00'))

#### Observations:
- This dataset has 6285 rows and 11 columns.
- All columns are there and 'date' column is in datetime format.
- Stocks included: YUM, BAC, GE, XOM, PFE
- Sectors included: Consumer Discretionary, Financials, Industrials, Energy, Healthcare.
- Date ranges from January 1, 2020 to December 31, 2024

## 4. Data Visualization

### 4.1 Trend Analysis  
We will:
- Visualize the closing price trends over time for each stock using line charts.
- Observe overall trends, growth, or declines during the selected time period.

### 4.2 Distribution & Variability  
We will:
- Use a histogram to visualize the distribution of daily returns across all stocks combined, identifying the overall shape, skewness, and presence of extreme returns.
- Use a boxplot to compare daily return variability by stock, highlighting which stocks show wider daily return spreads and more outliers.
- Use a boxplot of volatility by stock to compare long-term risk profiles, identifying which stocks exhibit higher sustained volatility levels over time.

### 4.3 Segment or Category Comparisons  
We will:
- Compare key risk and return metrics by sector to observe which sectors are generally more volatile or stable.
- Use bar charts to compare average daily return and average volatility by sector, providing a high-level view of sector performance and risk.

These comparisons will help us understand which sectors displayed higher resilience or risk during the 2020–2024 period, supporting broader investment insights.

### 4.4 COVID-19 Crash (March 2020) Analysis
We will:
- Analyze the stock performance during the initial COVID-19 crash (March 2020).
- Calculate the % change in closing price for each stock during March 2020.
- This offers insights into stock and sector resilience during periods of extreme market stress.


![Stock Closing Price Trends (2020-2024)](../visuals/closing_trends.png)

In [40]:
# Line chart displaying the closing price trends over time for each stock
plt.figure(figsize=(12,6))

# Plot each ticker
for ticker in df['Ticker'].unique():
    subset = df[df['Ticker'] == ticker]
    plt.plot(subset['Date'], subset['Close'], label=ticker)

plt.title('Stock Closing Price Trends (2020-2024)')
plt.xlabel('Date')
plt.ylabel('Closing Price')
plt.legend()
plt.savefig(r'C:\Users\Stuer\OneDrive\Professional\Stock_Performance_Volatility_Analysis\visuals\closing_trends.png', bbox_inches='tight')
plt.close()

![Distribution of Daily Returns (All Stocks)](../visuals/daily_return_histogram.png)

In [43]:
# Histogram of daily returns for all stocks combined
plt.figure(figsize=(10,6))
plt.hist(df['Daily_Return'], bins=50, edgecolor='black')
plt.title('Distribution of Daily Returns (All Stocks)')
plt.xlabel('Daily Return')
plt.ylabel('Frequency')
plt.savefig(r'C:\Users\Stuer\OneDrive\Professional\Stock_Performance_Volatility_Analysis\visuals\daily_return_histogram.png', bbox_inches='tight')
plt.close()

![Daily Return Variability by Stock](../visuals/daily_return_boxplot.png)

In [49]:
# Boxplot of daily return variability by stock
plt.figure(figsize=(10,6))
sns.boxplot(x='Ticker', y='Daily_Return', data=df)
plt.title('Daily Return Variability by Stock')
plt.xlabel('Stock')
plt.ylabel('Daily Return')
plt.savefig(r'C:\Users\Stuer\OneDrive\Professional\Stock_Performance_Volatility_Analysis\visuals\daily_return_boxplot.png', bbox_inches='tight')
plt.close()

![Volatility by Stock](../visuals/volatility_boxplot.png)

In [55]:
# Boxplot of volatility by stock
plt.figure(figsize=(10,6))
sns.boxplot(x='Ticker', y='Volatility', data=df)
plt.title('Volatility by Stock')
plt.xlabel('Stock')
plt.ylabel('Volatility')
plt.savefig(r'C:\Users\Stuer\OneDrive\Professional\Stock_Performance_Volatility_Analysis\visuals\volatility_boxplot.png', bbox_inches='tight')
plt.close()

![Average Daily Return by Sector](../visuals/avg_daily_return_by_sector.png)

In [57]:
# Bar chart of average daily return by sector
plt.figure(figsize=(8,6))
df.groupby('Sector')['Daily_Return'].mean().sort_values().plot(kind='bar')
plt.title('Average Daily Return by Sector (2020-2024)')
plt.xlabel('Sector')
plt.ylabel('Average Daily Return')
plt.xticks(rotation=45)
plt.savefig(r'C:\Users\Stuer\OneDrive\Professional\Stock_Performance_Volatility_Analysis\visuals\avg_daily_return_by_sector.png', bbox_inches='tight')
plt.close()

![Average Volatility by Sector](../visuals/avg_volatility_by_sector.png)

In [59]:
# Bar chart of average volatility by sector
plt.figure(figsize=(8,6))
df.groupby('Sector')['Volatility'].mean().sort_values().plot(kind='bar')
plt.title('Average Volatility by Sector (2020-2024)')
plt.xlabel('Sector')
plt.ylabel('Average Volatility')
plt.xticks(rotation=45)
plt.savefig(r'C:\Users\Stuer\OneDrive\Professional\Stock_Performance_Volatility_Analysis\visuals\avg_volatility_by_sector.png', bbox_inches='tight')
plt.close()

In [85]:
# Calculate % growth in closing price from first to last date per ticker
price_growth = df.groupby('Ticker')['Close'].agg(['first', 'last'])
price_growth['% Change'] = ((price_growth['last'] - price_growth['first']) / price_growth['first']) * 100
price_growth = price_growth.round(2)

print(price_growth)


         first    last  % Change
Ticker                          
BAC      35.64   43.91     23.20
GE       59.46  168.10    182.71
PFE      37.13   26.42    -28.85
XOM      70.90  105.76     49.17
YUM     102.17  133.52     30.68


In [87]:
# Calculate % change in closing price during COVID-19 crash (March 2020 only)
march_2020 = df[(df['Date'] >= '2020-03-01') & (df['Date'] <= '2020-03-31')]
march_growth = march_2020.groupby('Ticker')['Close'].agg(['first', 'last'])
march_growth['% Change'] = ((march_growth['last'] - march_growth['first']) / march_growth['first']) * 100
march_growth = march_growth.round(2)

print("\nCOVID-19 Crash Period Growth (March 2020):")
print(march_growth)


COVID-19 Crash Period Growth (March 2020):
        first   last  % Change
Ticker                        
BAC     29.37  21.23    -27.72
GE      55.87  39.57    -29.17
PFE     33.09  30.97     -6.42
XOM     53.88  37.97    -29.53
YUM     92.53  68.53    -25.94


## 5. Key Insights & Recommendations

| Observation            | Implication                               | Recommendation                                   |
|------------------------|-------------------------------------------|--------------------------------------------------|
| GE (+182.71%) and XOM (+49.17) showed the strongest price growth from 2020 to 2024.   | This suggests long-term growth, especially during later periods. | May be considered for growth-oriented portfolios if current momentum persists. |
| Healthcare had the lowest average daily return across all sectors.    | This sector underperformed compared to others over the analysis period. | This finding suggests the Healthcare sector may warrant closer review before considering any allocations. |
| Industrials and Energy sectors showed the highest average volatility  | These sectors are going to have more day-to-day risk and unpredictability. | Investors with lower risk tolerance might consider limiting exposure, particularly during periods of market volatility. |
| During the March 2020 COVID-19 crash, XOM (-29.53%), GE (-29.17%), and BAC (-27.72%) suffered the steepest declines. | Energy, Industrials, and Financials were highly sensitive to the COVID-19 market shock. | These sectors may be more vulnerable to macroeconomic shocks and require closer monitoring during global events. |
| PFE (-6.42%) was the most resilient stock during the March 2020 crash. | Healthcare showed defensive characteristics during the market downturn. | May be worth considering Healthcare stocks like PFE for diversification and downside protection in volatile periods. |

## 6. Limitations

- This analysis focuses only on historical stock data from 2020 to 2024 and does not incorporate forecasts or predictive modeling.
- Volatility was calculated using a 20-day rolling standard deviation, which captures short-term variability but may not fully represent all risk factors.
- The analysis does not include dividends, stock splits, or broader macroeconomic influences that may have impacted performance.
- The dataset was limited to a select group of stocks and sectors and may not reflect broader market dynamics.
- The COVID-19 analysis focused exclusively on the **initial crash phase (March 2020)** and did not examine the **entire pandemic period, including the recovery and post-crash performance**.